### PR TITLE
Alerting: Allow provenance disable in alerting provisioning API

### LINKED ISSUE #63047: Unable to edit alerts created from Provisioning API
**What happened**:
I recently upgraded Grafana from `v8.5.15` to `v9.3.2`. When I tried to edit the alerts, created using the Provisioning API, from the UI I found out that this is not currently possible and wanted to touch base to see what a possible solution could be.
 
I believe that updating an alert by PUT request with the whole JSON object is not as easy as changing things from the UI, especially when it comes to editing numerical values. This would significantly increase the chance of making an error thus creating problems in a production environment.
 
Could you please point me in the right direction?

**What you expected to happen**:
Being able to edit alerts from the UI like previous versions of Grafana.

**How to reproduce it (as minimally and precisely as possible)**:
Send a POST or PUT request with the Provisioning API with Grafana v9 or above.

**Anything else we need to know?**:

**Environment**:
- Grafana version: 9.3.2
- Data source type & version:
- OS Grafana is installed on: Ubuntu18.04
- User OS & Browser: MacOS and Chrome
- Grafana plugins:
- Others:

### PR DESCRIPTION
**Pull request by alexmobo**
**What is this feature?**

This PR allows us to disable provenance using `X-Disable-Provenance` headers in the following endpoints:

- `POST /api/v1/provisioning/alert-rules`
- `PUT /api/v1/provisioning/folder/{FolderUID}/rule-groups/{Group}` (calling this endpoint will change provenance for all alert rules within the alert group)
- `POST /api/v1/provisioning/contact-points`
- `POST /api/v1/provisioning/mute-timings`
- `PUT /api/v1/provisioning/policies`
- `PUT /api/v1/provisioning/templates/{name}`

**Why do we need this feature?**

Currently, disabling provenance in Alerting Provisioning API works when creating alert rules, but cannot be changed once created:

- `POST /api/v1/provisioning/alert-rules`

**Who is this feature for?**

All Alerting Provisioning API users that want to be able to edit those resources via UI.

**Which issue(s) does this PR fix?**:

Fixes https://github.com/grafana/grafana/issues/63047

**TESTS**:

All endpoints have been tested with Postman.

### CODE REVIEW

File: pkg/services/ngalert/provisioning/alert_rules_test.go [hunk 1]
```diff
@@ -213,10 +213,10 @@ func TestAlertRuleService(t *testing.T) {
 				errNil: false,
 			},
 			{
-				name:   "should not be able to update from provenance api to none",
+				name:   "should be able to update from provenance api to none",
 				from:   models.ProvenanceAPI,
 				to:     models.ProvenanceNone,
-				errNil: false,
+				errNil: true,
```
[the remark this answers is not in the document]
Comment (alexmobo) [reply pkg/services/ngalert/provisioning/alert_rules_test.go:219]: I understand what you say and I agree. I'll make the necessary changes. Thanks!

File: pkg/services/ngalert/provisioning/alert_rules.go [hunk 2]
```diff
@@ -271,13 +271,16 @@ func (service *AlertRuleService) ReplaceRuleGroup(ctx context.Context, orgID int
...
 		updates := make([]models.UpdateRule, 0, len(delta.Update))
 		for _, update := range delta.Update {
-			// check that provenance is not changed in a invalid way
+			// check that provenance is not changed in an invalid way
 			storedProvenance, err := service.provenanceStore.GetProvenance(ctx, update.New, orgID)
 			if err != nil {
 				return err
 			}
-			if storedProvenance != provenance && storedProvenance != models.ProvenanceNone {
+			err = canUpdateWithProvenance(storedProvenance, provenance, func() error {
```
Comment (alexmobo) [inline pkg/services/ngalert/provisioning/alert_rules.go:279]: I had to allow downgrading from provenance `API` to `None` in `ReplaceRuleGroup` as this method is a `replace`.

I understand this as if we removed all rules and inserted them again. So, this had to be done to allow insertions, updates, and deletes inside the method.

File: pkg/services/ngalert/provisioning/provenance.go [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+
+import (
+	"github.com/grafana/grafana/pkg/services/ngalert/models"
+)
+
+func canUpdateWithProvenance(storedProvenance, provenance models.Provenance, errFun func() error) error {
+	isFileProvenance := storedProvenance == models.ProvenanceFile
+	isAPItoFileProvenance := storedProvenance == models.ProvenanceAPI && provenance == models.ProvenanceFile
+	if storedProvenance != provenance && (isFileProvenance || isAPItoFileProvenance) {
+		return errFun()
```
Comment (JohnnyQQQQ) [inline pkg/services/ngalert/provisioning/provenance.go:11]: I don't get why we need a callback function here. Why are you not using error wrapping instead, which is a much more idiomatic go? 

i.e.:
```go
err := canUpdateWithProvenance(a, b)
if err != nil {
  return fmt.Errorf("something: %w", err)
}
```

File: pkg/services/ngalert/provisioning/provenance.go [hunk 4]
```diff
@@ -0,0 +1,14 @@
+package provisioning
+
+import (
+	"github.com/grafana/grafana/pkg/services/ngalert/models"
+)
+
+func canUpdateWithProvenance(storedProvenance, provenance models.Provenance, errFun func() error) error {
```
Comment (JohnnyQQQQ) [inline pkg/services/ngalert/provisioning/provenance.go:7]: Comment on how the function behaves. It's not apparent to the caller if it isn't a `bool` return type.

File: pkg/services/ngalert/provisioning/provenance.go (hunk 3 above)
Comment (alexmobo) [reply pkg/services/ngalert/provisioning/provenance.go]: I used a simpler approach. Thanks!

**Review by geek — APPROVED** [review]

**Comment by jantman** [comment]
Is there any update on this? Really hoping this will be merged and released soon...

**Review by JohnnyQQQQ — COMMENTED** [review]
Can we please add some documentation (https://grafana.com/docs/grafana/latest/alerting/set-up/provision-alerting-resources/view-provisioned-resources/) around this and how the semantics are? How does it work and what transitions are possible?

**Comment by grobinson-grafana** [comment]
> Can we please add some documentation (https://grafana.com/docs/grafana/latest/alerting/set-up/provision-alerting-resources/view-provisioned-resources/) around this and how the semantics are? How does it work and what transitions are possible?

This might be a better page https://grafana.com/docs/grafana/latest/alerting/set-up/provision-alerting-resources/file-provisioning/. What do you think JP?

**Review by brendamuir — APPROVED** [review]

**Comment by alexmobo** [comment]
> Can we please add some documentation (https://grafana.com/docs/grafana/latest/alerting/set-up/provision-alerting-resources/view-provisioned-resources/) around this and how the semantics are? How does it work and what transitions are possible?

Please, find the changed docs [here](https://github.com/grafana/grafana/pull/63650/commits/084e9e2cf9507690731dd0bd470868dce0b8d120)

**Review by grobinson-grafana — CHANGES_REQUESTED** [review]
I'm not 100% sure what is supposed to happen. `X-Disable-Provenance` works for `POST` in `main`, shouldn't it now work for `PUT`? I just tried this for an existing alert rule and got the following error, but looking at the `canUpdateProvenanceInRuleGroup` function I expected it to work:

```
{"message":"cannot change provenance from 'api' to ''","traceID":""}
```

I'm a little confused about what I'm supposed to be testing to make sure this works.

File: pkg/services/ngalert/provisioning/provenance.go [hunk 5]
```diff
@@ -0,0 +1,13 @@
+package provisioning
+
+import (
+	"github.com/grafana/grafana/pkg/services/ngalert/models"
+)
+
+// canUpdateProvenanceInRuleGroup checks if a provenance can be updated for a rule group and its alerts.
```
[the remark this answers is not in the document]
Comment (alexmobo) [reply pkg/services/ngalert/provisioning/provenance.go:7]: Yes, that's it.

File: pkg/services/ngalert/provisioning/alert_rules_test.go [hunk 6]
```diff
@@ -274,10 +274,10 @@ func TestAlertRuleService(t *testing.T) {
 				errNil: false,
 			},
 			{
-				name:   "should not be able to update from provenance api to none",
+				name:   "should be able to update from provenance api to none",
```
[the remark this answers is not in the document]
Comment (alexmobo) [reply pkg/services/ngalert/provisioning/alert_rules_test.go:277]: As per @alexweav comment https://github.com/grafana/grafana/pull/63650#discussion_r1126744297, I understood we only have to change the semantics in regards with the issue we are trying to solve, which is https://github.com/grafana/grafana/issues/63047.

That means that alert rules are the only ones that can change from `api` to `none` as that is the requirement in that issue.

**Comment by alexmobo** [comment]
@brendamuir can you please check these changes https://github.com/grafana/grafana/pull/63650/commits/f78729f5b4e4dde1e15c4fe74075d5bb8460be61?

File: docs/sources/alerting/set-up/provision-alerting-resources/_index.md [hunk 7]
```diff
@@ -24,21 +24,23 @@ There are three options to choose from:
...
 
-   PUT /api/v1/provisioning/alert-rules/{UID}
+- `POST /api/v1/provisioning/alert-rules`
+- `POST /api/v1/provisioning/contact-points`
+- `POST /api/v1/provisioning/mute-timings`
+- `PUT /api/v1/provisioning/policies`
+- `PUT /api/v1/provisioning/templates/{name}`
 
-1. Provision your alerting resources using Terraform.
+The next API call will change provenance of all alert rules in the group.
```
[the remark this answers is not in the document]
Comment (alexmobo) [reply docs/sources/alerting/set-up/provision-alerting-resources/_index.md]: Made a change for that https://github.com/grafana/grafana/pull/63650/commits/f78729f5b4e4dde1e15c4fe74075d5bb8460be61

**Review by grobinson-grafana — APPROVED** [review]

**Comment by grobinson-grafana** [comment]
I've updated the description to use the endpoints from the documentation.

**Review by brendamuir — APPROVED** [review]

**Pull request merged**